system_zero/src/alu/bitops.rs

```diff
@@ -134,28 +118,40 @@ pub(crate) fn eval_bitop<F: Field, P: PackedField<Scalar = F>>(
         COL_BIT_DECOMP_INPUT_A_LO_BIN_REGS,
         COL_BIT_DECOMP_INPUT_B_LO_BIN_REGS,
         COL_BITOP_OUTPUT_0,
-        COL_BITOP_OUTPUT_1,
         yield_constr,
     );
     // Constraint for hi half
     eval_bitop_32(
         lv,
         COL_BIT_DECOMP_INPUT_A_HI_BIN_REGS,
         COL_BIT_DECOMP_INPUT_B_HI_BIN_REGS,
-        COL_BITOP_OUTPUT_2,
-        COL_BITOP_OUTPUT_3,
+        COL_BITOP_OUTPUT_1,
         yield_constr,
     );
 }
 
+fn constrain_all_to_bits<F: RichField + Extendable<D>, const D: usize>(
+    builder: &mut CircuitBuilder<F, D>,
+    yield_constr: &mut RecursiveConstraintConsumer<F, D>,
+    filter: ExtensionTarget<D>,
+    vals: &[ExtensionTarget<D>]
+) {
+    for v in vals.into_iter() {
+        let one = builder.one_extension();
+        let t0 = builder.sub_extension(one, *v);
```

Comment:
Normally this would be too minor to bother, but since we'll have 128 iterations here, I think it'd be good to change `v * (1 - v)` to `v * v - v`, so we can use `mul_sub(v, v, v)` to save one arithmetic op.

(Since that switches `1 - v` to `v - 1`, I think the same change would need to be made in `eval_bitop_32`.)

Reply:
Good catch!